asadiqbal08/mitxpro - CMS: customize certificates

asadiqbal08 opened the pull request:
Task: #872 

--------
### SCREEN SHOTS
--------
#### CERTIFICATE LOOK & FEEL
![screencapture-xpro-odl-local-8053-courses-course-v1-xPRO-DgtlLearn1-certificate-14-2019-07-31-16_21_54](https://user-images.githubusercontent.com/7334669/62208733-243be880-b3b1-11e9-9c7e-9cca9093588f.png)

#### MANAGE SIGNATORIES SEPARATELY AND MOVED UNDER SIGNATORY FOLDER OVER CMS. 
![Screen Shot 2019-07-26 at 6 16 25 PM](https://user-images.githubusercontent.com/7334669/61954507-6be9fb00-afd2-11e9-8361-ad1ce062b0ef.jpg)

#### SIGNATORIES FIELDS. 
![screencapture-xpro-odl-local-8053-cms-pages-add-cms-signatorypage-102-2019-07-31-16_31_22](https://user-images.githubusercontent.com/7334669/62208759-36b62200-b3b1-11e9-8f95-705262114644.png)

#### SIGNATORIES ARE LISTED SEPARATELY. 

![Screen Shot 2019-07-31 at 4 20 52 PM](https://user-images.githubusercontent.com/7334669/62208775-4170b700-b3b1-11e9-9b93-b91e5fb7ad68.png)

#### CERTIFICATES CAN BE CREATED UNDER COURSE/PROGRAM PAGES.
<img width="1117" alt="Screen Shot 2019-07-26 at 6 16 55 PM" src="https://user-images.githubusercontent.com/7334669/61954515-72787280-afd2-11e9-852d-019feb56ab69.png">

#### CAN ASSOCIATE UPTO 5 SIGNATORIES.
![Screen Shot 2019-07-31 at 4 21 28 PM](https://user-images.githubusercontent.com/7334669/62208801-49c8f200-b3b1-11e9-8e24-b19f6e51a264.png)

#### Certificate Page Title

![Screen Shot 2019-07-31 at 4 21 36 PM](https://user-images.githubusercontent.com/7334669/62208826-5a796800-b3b1-11e9-8187-ed6c1d450229.png)


#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots
  - [ ] Mobile width screenshots
  - [ ] Tag @ferdi or @pdpinch for review
- [ ] Migrations
  - [ ] Migration is backwards-compatible with current production code
- [ ] Testing
  - [ ] Code is tested
  - [ ] Changes have been manually tested
- [ ] Settings
  - [ ] New settings are documented and present in `app.json`
  - [ ] New settings have reasonable development defaults, if applicable

asadiqbal08 commented:
@ahmed-arbisoft @pdpinch please feel free to share your thoughts over it. 

please do not go through the actual code review but should look around wagtail page design and code structures.

ahmed-arbisoft commented:
I'm guessing signatories will be sort of like:
John Doe, Associate Professor, XYZ
Jane Doe, Professor, ABC

Could you rename the `title` to `name` and remove the previous name field? Because these are people, name sounds more fitting. Also, do we need separate `title_1` and `title_2`? Isn't one title enough?

For `CertificatePage` I don't think `name` is a valid field. The names on the certificate would be coming from either the signatories or the learner themselves.

asadiqbal08 commented:
Thanks @ahmed-arbisoft for the initial feedback but I followed the things as per A/C in #872. 

> Could you rename the title to name and remove the previous name field? Because these are people, name sounds more fitting. Also, do we need separate title_1 and title_2? Isn't one title enough?

If you look into the image below, I tried to explain what is `Name`, `title 1`, `title 2` : 
![explain](https://user-images.githubusercontent.com/7334669/61970847-11af6100-aff7-11e9-8b08-5062e007e425.jpg)
If user specify the `organization` then it will appear below `title 2`. 
So it seems me, `title 1`, `title 2` and `name` are different fields in case of signatory 

> For CertificatePage I don't think name is a valid field. The names on the certificate would be coming from either the signatories or the learner themselves.

I added the `Name` ( I will mark it `not required` later), sure it will be the name of the `user` but I just added here for CM team. If someone wants to preview the name over certificate.

pdpinch commented:
For the signatory, I would expect `Name` to be the first field, and the only required one. 

For the certificate, can we change the label for "Title" to "Course/Program Title". The "Name" field is going to be confusing. Do we need it? Can we change the "CEU" label to "CEUs"?

asadiqbal08 commented:
> For the signatory, I would expect Name to be the first field, and the only required one.
Yes, other will be optional.

> For the certificate, can we change the label for "Title" to "Course/Program Title". 
we are using the wagtail page default title field as doing in other pages. `Title` here is representing certificate title.  

> The "Name" field is going to be confusing. Do we need it? 
It is optional for previewing user name. I can remove it or either change it to user_name etc. 
> Can we change the "CEU" label to "CEUs"?
Yes.

ahmed-arbisoft commented on cms/templates/certificate_page.html:
```diff
@@ -0,0 +1,46 @@
+            <div class="institute-logo">
+                <img src="{% static 'images/certificates/certificate.png' %}" alt="">
+            </div>
+            <span class="certify-text">This is certify that</span>
```
This is _to_ certify that

ahmed-arbisoft commented on static/scss/certificates/certificate.scss:
```diff
@@ -0,0 +1,123 @@
+    display: block;
+    font-size: 18px;
+    line-height: 25px;
+    color: #6b6969;
```
Can we extract this color into a variable?

asadiqbal08 replied:
I was thinking but @ahmed-arbisoft truly speaking the current `variables.scss` is really a confusing file. personally do not prefer to add there.

ahmed-arbisoft replied:
No I meant just declare this variable at the top of this file, that would be fine as well.

asadiqbal08 commented on cms/models.py:
```diff
@@ -1069,6 +1111,149 @@ def get_context(self, request, *args, **kwargs):
+        """
+        We need to serve the certificate template for preview.
+        """
+        return Page.serve(self, request, *args, **kwargs)
```
I remember, I tried it before too but getting `'super' object has no attribute 'serve'`.

ahmed-arbisoft replied:
Yeah I guess it's okay then.

ahmed-arbisoft commented on cms/models.py:
```diff
@@ -1071,6 +1113,149 @@ def get_context(self, request, *args, **kwargs):
+        help_text="Specify the organization of signatory.",
+    )
+
+    image = models.ForeignKey(
```
Can we just go ahead and rename this to `signature_image`? It will remove any chance of a confusion.

asadiqbal08 replied:
sure.

ahmed-arbisoft commented on cms/models.py:
```diff
@@ -1071,6 +1113,149 @@ def get_context(self, request, *args, **kwargs):
+        if not self.title:
+            self.title = self.__class__._meta.verbose_name.title() + "-" + self.name
+
+        self.slug = slugify("{}-{}-{}".format(self.title, self.name, self.id))
```
`self.title` is already `{_meta.verbose_name.title}-{self.name}` so the slug will become `{verbose_name.title}-{self.name}-{self.name}-{self.id}`. Might want to remove `self.name` from `self.slug`.

ahmed-arbisoft commented on cms/templates/certificate_page.html:
```diff
@@ -0,0 +1,46 @@
+            </div>
+            <span class="institute-text">Massachusetts Institute of Technology</span>
+            <div class="institute-logo">
+                <img src="{% static 'images/certificates/certificate.png' %}" alt="">
```
nit: How about `certificates/logo.png`?

ahmed-arbisoft commented on cms/models.py:
```diff
@@ -1069,6 +1111,149 @@ def get_context(self, request, *args, **kwargs):
+        return self.get_parent().specific
+
+    def get_context(self, request, *args, **kwargs):
+        return {
```
You can override the request or simply pass an extra `is_preview` parameter by implementing the `serve_preview` method for this class.

asadiqbal08 replied:
@ahmed-arbisoft we are not using the `preview` button but the `live` one. I believe, `is_preview` is relates to `preview`

asadiqbal08 commented on cms/models.py:
```diff
@@ -1248,11 +1248,19 @@ def parent(self):
+        preview_context = {}
+
+        if request.is_preview:
+            preview_context = {
```
@ahmed-arbisoft I have added a `preview_context` here that will render the certificate view version while clicking on `PREVIEW` button. 
Ref: 

<img width="1120" alt="Screen Shot 2019-08-01 at 4 57 38 PM" src="https://user-images.githubusercontent.com/7334669/62291490-806c3e80-b47d-11e9-93db-13976a6ea316.png">

pdpinch commented:
How do we set the course name as it appears on the certificate?

asadiqbal08 commented:
> How do we set the course name as it appears on the certificate?

I am picking the associated parent course/program page title [here](https://github.com/mitodl/mitxpro/pull/911/files#diff-c017effe5d76e267a79bcadafa31b4d8R18)

pdpinch commented:
Sorry, I'm still not clear on the course name on the certificate. It needs to be customizable, i.e. we may use one name on the product page in the catalog, and another name on the certificate. How do I do that? And how do I edit it after I've customized it?

asadiqbal08 commented:
added `Product Name` in certificatepage. Going to merge after build pass.
